File: android/app/src/main/java/com/mengzhen/app/scheduler/TaskStorage.java
package com.mengzhen.app.scheduler;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * 持久化任务信息到 SharedPreferences，设备重启后可恢复
 */
public class TaskStorage {

    private static final String PREFS_NAME = "alarm_tasks";
    private static TaskStorage instance;
    private final Context context;

    private TaskStorage(Context context) {
        this.context = context.getApplicationContext();
    }

    public static synchronized TaskStorage getInstance(Context context) {
        if (instance == null) {
            instance = new TaskStorage(context);
        }
        return instance;
    }

    public void saveTask(TaskInfo task) {
        try {
            JSONArray arr = getAllJsonArray();
            // 先删除同 id 的旧任务
            JSONArray filtered = new JSONArray();
            for (int i = 0; i < arr.length(); i++) {
                JSONObject obj = arr.getJSONObject(i);
                if (!obj.getString("taskId").equals(task.taskId)) {
                    filtered.put(obj);
                }
            }
            filtered.put(task.toJson());
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                    .edit()
                    .putString("tasks", filtered.toString())
                    .apply();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void removeTask(String taskId) {
        try {
            JSONArray arr = getAllJsonArray();
            JSONArray filtered = new JSONArray();
            for (int i = 0; i < arr.length(); i++) {
                JSONObject obj = arr.getJSONObject(i);
                if (!obj.getString("taskId").equals(taskId)) {
                    filtered.put(obj);
                }
            }
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                    .edit()
                    .putString("tasks", filtered.toString())
                    .apply();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public List<TaskInfo> getAllTasks() {
        List<TaskInfo> list = new ArrayList<>();
        try {
            JSONArray arr = getAllJsonArray();
            for (int i = 0; i < arr.length(); i++) {
                JSONObject obj = arr.getJSONObject(i);
                list.add(TaskInfo.fromJson(obj));
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return list;
    }

    public TaskInfo getTask(String taskId) {
        for (TaskInfo task : getAllTasks()) {
            if (task.taskId.equals(taskId)) return task;
        }
        return null;
    }

    private JSONArray getAllJsonArray() {
        String json = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                .getString("tasks", "[]");
        try {
            return new JSONArray(json);
        } catch (Exception e) {
            return new JSONArray();
        }
    }
}

File: android/app/src/main/java/com/mengzhen/app/scheduler/TaskStorage.kt
package com.mengzhen.app.scheduler

import android.content.Context
import org.json.JSONArray

class TaskStorage private constructor(context: Context) {

    private val prefs = context.getSharedPreferences("dream_pillow_alarms", Context.MODE_PRIVATE)

    fun saveTask(task: TaskInfo) {
        val tasks = getAllTasks().filterNot { it.taskId == task.taskId }.toMutableList()
        tasks.add(task)
        persist(tasks)
    }

    fun removeTask(taskId: String) {
        persist(getAllTasks().filterNot { it.taskId == taskId })
    }

    fun getTask(taskId: String): TaskInfo? = getAllTasks().find { it.taskId == taskId }

    fun getAllTasks(): List<TaskInfo> {
        val raw = prefs.getString(KEY, null) ?: return emptyList()
        return try {
            val arr = JSONArray(raw)
            (0 until arr.length()).map { taskInfoFromJson(arr.getJSONObject(it)) }
        } catch (e: Exception) {
            emptyList()
        }
    }

    private fun persist(tasks: List<TaskInfo>) {
        val arr = JSONArray()
        tasks.forEach { arr.put(it.toJson()) }
        prefs.edit().putString(KEY, arr.toString()).apply()
    }

    /** 清理已完成的一次性任务 - 对标 Web 端 cleanupCompletedOnceTasks */
    fun cleanupCompletedOnceTasks(): Int {
        val tasks = getAllTasks().toMutableList()
        val before = tasks.size
        tasks.removeAll { it.status == "completed" && it.repeatType == 0 && it.repeatDays == 0 }
        val removed = before - tasks.size
        if (removed > 0) persist(tasks)
        return removed
    }

    /** 清理已取消的任务 - 对标 Web 端 cleanupCancelledTasks */
    fun cleanupCancelledTasks(): Int {
        val tasks = getAllTasks().toMutableList()
        val before = tasks.size
        tasks.removeAll { it.status == "cancelled" }
        val removed = before - tasks.size
        if (removed > 0) persist(tasks)
        return removed
    }

    companion object {
        private const val KEY = "alarm_tasks"
        @Volatile private var instance: TaskStorage? = null
        fun get(context: Context): TaskStorage =
            instance ?: synchronized(this) {
                instance ?: TaskStorage(context.applicationContext).also { instance = it }
            }
    }
}
